lib/spack/spack/config.py

```diff
@@ -944,6 +944,10 @@ def they_are(t):
         # track keys for marking
         key_marks = {}
 
+        # track this to ensure that source items come *before* dest in
+        # iteration order with OrderdDicts
+        dest_keys = [dk for dk in dest.keys() if dk not in source]
```

Comment:
if dk not in source` may be worth mentioning in the docstring: technically that's covered by "entries in source take precedence over dest", but you could also mention that keys in both will have their order updated to match the order of keys in the source.

Reply:
I think this is addressed.